File: src/lib/db/quotes.ts
/** Quote requests: anyone can send one, administrators work through them. */

import "server-only";
import { one, query, scalar } from "./client";
import { ensureReady } from "./init";
import type { Page, QuoteLine, QuoteRow, QuoteStatus } from "./types";

export interface QuoteInput {
  contact_name: string;
  contact_email: string;
  organisation?: string | null;
  phone?: string | null;
  city?: string | null;
  timeline?: string | null;
  target?: string;
  workloads?: string[];
  notes?: string | null;
  lines?: QuoteLine[];
  summary?: Record<string, unknown>;
  findings?: Array<Record<string, unknown>>;
  submission_hash?: string;
  subtotal_pkr?: number;
}

export async function submitQuote(q: QuoteInput): Promise<QuoteRow> {
  await ensureReady();
  const email = q.contact_email.toLowerCase();

  // Linked to an account when the email matches one, so it shows up in their
  // history. Deliberately not required: nobody should need an account to ask
  // what something costs.
  const userId = await scalar<number>("SELECT id FROM users WHERE email = $1", [email]);
  const customer = await one<{ id: number }>(`INSERT INTO customers(customer_number,organisation,display_name,email,phone)
    VALUES(next_document_number('customer','CUS'),$1,$2,$3,$4)
    ON CONFLICT (lower(email)) WHERE email IS NOT NULL DO UPDATE SET
      organisation=COALESCE(EXCLUDED.organisation,customers.organisation), display_name=EXCLUDED.display_name,
      phone=COALESCE(EXCLUDED.phone,customers.phone), updated_at=now() RETURNING id`,
    [q.organisation ?? null,q.contact_name,email,q.phone ?? null]);
  const reference = await scalar<string>("SELECT next_document_number('quote','SCQ')");

  const rows = await query<QuoteRow>(
    `INSERT INTO quotes
       (reference, user_id, customer_id, contact_name, contact_email, organisation, phone, city,
        timeline, target, workloads, notes, lines, summary, findings, submission_hash, subtotal_pkr)
     VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16,$17)
     RETURNING *`,
    [
      reference, userId, customer?.id ?? null, q.contact_name, email, q.organisation ?? null,
      q.phone ?? null, q.city ?? null, q.timeline ?? null, q.target ?? "desk",
      JSON.stringify(q.workloads ?? []), q.notes ?? null,
      JSON.stringify(q.lines ?? []), JSON.stringify(q.summary ?? {}),
      JSON.stringify(q.findings ?? []), q.submission_hash ?? null, q.subtotal_pkr ?? 0,
    ],
  );
  return rows[0];
}

export async function createQuoteRevision(reference:string,actorId:number,note?:string):Promise<number>{
  await ensureReady();
  const quote=await getQuote(reference); if(!quote) throw new Error("Quote not found.");
  const revision=Number(quote.revision_number)+1;
  await query(`INSERT INTO quote_revisions(quote_id,revision_number,lines,summary,findings,subtotal_pkr,tax_rate,discount_pkr,valid_until,payment_terms,note,created_by)
    VALUES($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12)`,[quote.id,revision,JSON.stringify(quote.lines),JSON.stringify(quote.summary),JSON.stringify(quote.findings),quote.subtotal_pkr,quote.tax_rate,quote.discount_pkr,quote.valid_until,quote.payment_terms,note??null,actorId]);
  await query("UPDATE quotes SET revision_number=$1,status='reviewing',updated_at=now() WHERE id=$2",[revision,quote.id]);
  return revision;
}

export async function quoteRevisions(reference:string){await ensureReady();return query<{revision_number:number;note:string|null;created_at:Date;actor_email:string|null}>(`SELECT r.revision_number,r.note,r.created_at,u.email actor_email FROM quote_revisions r JOIN quotes q ON q.id=r.quote_id LEFT JOIN users u ON u.id=r.created_by WHERE q.reference=$1 ORDER BY r.revision_number DESC`,[reference]);}

export async function recentQuoteByHash(hash: string): Promise<QuoteRow | null> {
  await ensureReady();
  return one<QuoteRow>(
    `SELECT * FROM quotes
      WHERE submission_hash = $1 AND created_at > now() - interval '10 minutes'
      ORDER BY created_at DESC LIMIT 1`,
    [hash],
  );
}

export async function listQuotes(
  opts: { status?: string; page?: number; perPage?: number } = {},
): Promise<Page<QuoteRow>> {
  await ensureReady();

  const params: unknown[] = [];
  let clause = "";
  if (opts.status) {
    params.push(opts.status);
    clause = `WHERE status = $${params.length}`;
  }

  const total = Number(await scalar<string>(`SELECT count(*) FROM quotes ${clause}`, params));
  const perPage = Math.min(Math.max(opts.perPage ?? 25, 1), 100);
  const page = Math.max(opts.page ?? 1, 1);

  const items = await query<QuoteRow>(
    `SELECT * FROM quotes ${clause} ORDER BY created_at DESC
      LIMIT $${params.length + 1} OFFSET $${params.length + 2}`,
    [...params, perPage, (page - 1) * perPage],
  );

  return { items, total, page, pages: Math.max(1, Math.ceil(total / perPage)) };
}

export async function getQuote(reference: string): Promise<QuoteRow | null> {
  await ensureReady();
  return one<QuoteRow>("SELECT * FROM quotes WHERE reference = $1", [reference]);
}

export async function updateQuote(
  reference: string,
  patch: {
    status?: QuoteStatus;
    internal_note?: string | null;
    lines?: QuoteLine[];
    subtotal_pkr?: number;
    tax_rate?: number;
    discount_pkr?: number;
    valid_until?: string | null;
    payment_terms?: string | null;
    sent_at?: Date | null;
  },
): Promise<void> {
  await ensureReady();

  const sets: string[] = [];
  const params: unknown[] = [];
  if (patch.status !== undefined) {
    params.push(patch.status);
    sets.push(`status = $${params.length}`);
  }
  if (patch.internal_note !== undefined) {
    params.push(patch.internal_note);
    sets.push(`internal_note = $${params.length}`);
  }
  if (patch.lines !== undefined) {
    params.push(JSON.stringify(patch.lines));
    sets.push(`lines = $${params.length}`);
  }
  for (const key of ["subtotal_pkr", "tax_rate", "discount_pkr"] as const) {
    if (patch[key] !== undefined) {
      params.push(patch[key]);
      sets.push(`${key} = $${params.length}`);
    }
  }
  if (patch.valid_until !== undefined) {
    params.push(patch.valid_until || null);
    sets.push(`valid_until = $${params.length}`);
  }
  if (patch.payment_terms !== undefined) {
    params.push(patch.payment_terms);
    sets.push(`payment_terms = $${params.length}`);
  }
  if (patch.sent_at !== undefined) { params.push(patch.sent_at); sets.push(`sent_at = $${params.length}`); }
  if (!sets.length) return;

  params.push(reference);
  await query(`UPDATE quotes SET ${sets.join(", ")}, updated_at = now() WHERE reference = $${params.length}`, params);
}

/**
 * A signed-in customer's own requests.
 *
 * Matched on the email as well as the account id, so a request sent before they
 * registered still appears once they sign up with that address.
 */
export async function quotesForUser(userId: number, email: string): Promise<QuoteRow[]> {
  await ensureReady();
  return query<QuoteRow>(
    `SELECT * FROM quotes WHERE user_id = $1 OR contact_email = $2
      ORDER BY created_at DESC LIMIT 50`,
    [userId, email.toLowerCase()],
  );
}

export interface Stats {
  productsTotal: number;
  productsActive: number;
  productsInStock: number;
  quotesTotal: number;
  quotesNew: number;
  usersTotal: number;
  invoicesTotal: number;
  invoicesOutstanding: number;
}

export async function stats(): Promise<Stats> {
  await ensureReady();
  // One trip rather than six.
  const row = await one<Record<string, string>>(`
    SELECT
      (SELECT count(*) FROM products)                        AS products_total,
      (SELECT count(*) FROM products WHERE is_active)        AS products_active,
      (SELECT count(*) FROM products WHERE stock_qty > 0)    AS products_in_stock,
      (SELECT count(*) FROM quotes)                          AS quotes_total,
      (SELECT count(*) FROM quotes WHERE status = 'new')     AS quotes_new,
      (SELECT count(*) FROM users)                           AS users_total,
      (SELECT count(*) FROM invoices)                        AS invoices_total,
      (SELECT count(*) FROM invoices WHERE status IN ('draft','sent')) AS invoices_outstanding
  `);
  return {
    productsTotal: Number(row?.products_total ?? 0),
    productsActive: Number(row?.products_active ?? 0),
    productsInStock: Number(row?.products_in_stock ?? 0),
    quotesTotal: Number(row?.quotes_total ?? 0),
    quotesNew: Number(row?.quotes_new ?? 0),
    usersTotal: Number(row?.users_total ?? 0),
    invoicesTotal: Number(row?.invoices_total ?? 0),
    invoicesOutstanding: Number(row?.invoices_outstanding ?? 0),
  };
}

File: backend/app/routers/quotes.py
"""Quote requests: anyone can submit one, administrators work through them."""

from datetime import datetime, timezone
from math import ceil
from secrets import token_hex
from typing import Annotated

from fastapi import APIRouter, HTTPException, Query, status
from sqlalchemy import func, or_, select

from ..auth import CurrentAdmin, CurrentUser, DbSession
from ..models import Product, Quote, QuoteStatus, User
from ..schemas import QuoteCreate, QuoteOut, QuotePage, QuoteUpdate, Stats

public = APIRouter(prefix="/api/quotes", tags=["quotes"])
admin = APIRouter(prefix="/api/admin", tags=["admin"])


def make_reference() -> str:
    stamp = datetime.now(timezone.utc).strftime("%Y%m%d")
    return f"SC-{stamp}-{token_hex(2).upper()}"


@public.post("", response_model=QuoteOut, status_code=status.HTTP_201_CREATED)
def submit_quote(body: QuoteCreate, db: DbSession) -> QuoteOut:
    """Accept a quote request.

    Deliberately open: a customer should not need an account to ask for a price.
    If the email matches a registered user the quote is linked to them so it
    appears in their history, but that is a convenience, not a requirement.
    """
    user = db.query(User).filter(User.email == body.contact_email.lower()).first()

    quote = Quote(
        reference=make_reference(),
        user_id=user.id if user else None,
        contact_name=body.contact_name,
        contact_email=body.contact_email.lower(),
        organisation=body.organisation,
        phone=body.phone,
        city=body.city,
        timeline=body.timeline,
        target=body.target,
        workloads=body.workloads,
        notes=body.notes,
        lines=[line.model_dump() for line in body.lines],
        summary=body.summary,
        findings=body.findings,
    )
    db.add(quote)
    db.commit()
    db.refresh(quote)
    return QuoteOut.model_validate(quote)


@public.get("/mine", response_model=list[QuoteOut])
def my_quotes(user: CurrentUser, db: DbSession) -> list[QuoteOut]:
    """The signed-in user's own quote requests, newest first.

    Matched on the email as well as the account id, so a request sent before
    the account existed still shows up once they register with that address.
    """
    rows = db.scalars(
        select(Quote)
        .where(or_(Quote.user_id == user.id, Quote.contact_email == user.email))
        .order_by(Quote.created_at.desc())
        .limit(50)
    ).all()
    return [QuoteOut.model_validate(r) for r in rows]


@admin.get("/quotes", response_model=QuotePage)
def list_quotes(
    db: DbSession,
    _: CurrentAdmin,
    status_filter: str | None = Query(default=None, alias="status"),
    q: str | None = None,
    page: Annotated[int, Query(ge=1)] = 1,
    per_page: Annotated[int, Query(ge=1, le=100)] = 25,
) -> QuotePage:
    stmt = select(Quote)
    if status_filter:
        stmt = stmt.where(Quote.status == status_filter)
    if q:
        like = f"%{q.lower()}%"
        stmt = stmt.where(
            or_(
                func.lower(Quote.reference).like(like),
                func.lower(Quote.contact_name).like(like),
                func.lower(Quote.contact_email).like(like),
                func.lower(func.coalesce(Quote.organisation, "")).like(like),
            )
        )
    stmt = stmt.order_by(Quote.created_at.desc())

    total = db.scalar(stmt.with_only_columns(func.count(Quote.id)).order_by(None)) or 0
    rows = db.scalars(stmt.limit(per_page).offset((page - 1) * per_page)).all()
    return QuotePage(
        items=[QuoteOut.model_validate(r) for r in rows],
        total=total,
        page=page,
        pages=max(1, ceil(total / per_page)),
    )


@admin.get("/quotes/{reference}", response_model=QuoteOut)
def get_quote(reference: str, db: DbSession, _: CurrentAdmin) -> QuoteOut:
    row = db.query(Quote).filter(Quote.reference == reference).first()
    if row is None:
        raise HTTPException(status_code=404, detail="No such quote.")
    return QuoteOut.model_validate(row)


@admin.patch("/quotes/{reference}", response_model=QuoteOut)
def update_quote(reference: str, body: QuoteUpdate, db: DbSession, _: CurrentAdmin) -> QuoteOut:
    row = db.query(Quote).filter(Quote.reference == reference).first()
    if row is None:
        raise HTTPException(status_code=404, detail="No such quote.")
    for field, value in body.model_dump(exclude_unset=True).items():
        setattr(row, field, value)
    db.commit()
    db.refresh(row)
    return QuoteOut.model_validate(row)


@admin.get("/stats", response_model=Stats)
def stats(db: DbSession, _: CurrentAdmin) -> Stats:
    by_kind = dict(
        db.execute(
            select(Product.kind, func.count(Product.id))
            .where(Product.is_active.is_(True))
            .group_by(Product.kind)
            .order_by(func.count(Product.id).desc())
        ).all()
    )
    return Stats(
        products_total=db.scalar(select(func.count(Product.id))) or 0,
        products_active=db.scalar(select(func.count(Product.id)).where(Product.is_active.is_(True))) or 0,
        products_in_stock=db.scalar(select(func.count(Product.id)).where(Product.stock_qty > 0)) or 0,
        quotes_total=db.scalar(select(func.count(Quote.id))) or 0,
        quotes_new=db.scalar(select(func.count(Quote.id)).where(Quote.status == QuoteStatus.new)) or 0,
        users_total=db.scalar(select(func.count(User.id))) or 0,
        by_kind=by_kind,
    )
